src/libcollections/slice.rs

```diff
@@ -1553,28 +1661,20 @@ unsafe fn merge<T, F>(v: &mut [T], mid: usize, buf: *mut T, is_less: &mut F)
 fn merge_sort<T, F>(v: &mut [T], mut is_less: F)
     where F: FnMut(&T, &T) -> bool
 {
+    // Slices of up to this length get sorted using insertion sort.
+    const MAX_INSERTION: usize = 16;
+    // Very short runs are extended using insertion sort to span at least this many elements.
+    const MIN_RUN: usize = 8;
+
     // Sorting has no meaningful behavior on zero-sized types.
     if size_of::<T>() == 0 {
         return;
     }
 
-    // FIXME #12092: These numbers are platform-specific and need more extensive testing/tuning.
-    //
-    // If `v` has length up to `max_insertion`, simply switch to insertion sort because it is going
-    // to perform better than merge sort. For bigger types `T`, the threshold is smaller.
-    //
-    // Short runs are extended using insertion sort to span at least `min_run` elements, in order
-    // to improve performance.
```

Comment:
I don't believe it's worthwhile differentiating between big and small `T`s - it doesn't have a large effect.

Instead, it's better to focus on some middle ground that works well all around: big/small `T`s and expensive/cheap comparison functions as well. I took a look at a `std::stable_sort` implementation in C++ and it has similar constants.